### PR TITLE
Add List.combine_chop and corresponding (and&) synchronized product

### PR DESCRIPTION
**Pull request by grayswandyr**
Implements #335.
Sorry, the former PR about `mguard` is also inside the PR and already pushed... Hope you can sort this out.

Also : I have only implemented the function for List for the time being. Indeed, it seems neither CCSeq nor OSeq come with a let syntax, and `zip` is already implemented in OSeq.

### CODE REVIEW

File: src/core/CCList.ml [hunk 1]
```diff
@@ -570,6 +572,29 @@ let combine_gen l1 l2 =
     res1 = res2)
 *)
 
+let combine_chop l1 l2 =
```
Comment (c-cube) [inline src/core/CCList.ml:575]:
```suggestion
let combine_shortest l1 l2 =
```

File: src/core/CCList.ml [hunk 2]
```diff
@@ -570,6 +572,29 @@ let combine_gen l1 l2 =
...
+  let rec direct i l1 l2 = match l1, l2 with
+    | (_, []) | ([], _) -> []
+    | _ when i=0 -> safe l1 l2 []
+    | (x1::l1', x2::l2') -> (x1, x2) :: direct (i-1) l1' l2'
+  and safe l1 l2 acc = match l1, l2 with
+    | ([], _) | (_, []) -> List.rev acc
+    | (x1::l1', x2::l2') -> safe l1' l2' @@ (x1, x2) :: acc
```
Comment (c-cube) [inline src/core/CCList.ml:582]:
```suggestion
    | (x1::l1', x2::l2') ->
      let acc = (x1,x2) :: acc in
      safe l1' l2' acc
```

(clearer what evaluation order is, and avoids the `@@`)

File: src/core/CCList.ml [hunk 3]
```diff
@@ -1784,6 +1809,8 @@ module Infix = struct
       let (>>=) = (>>=)
       let[@inline]  monoid_product l1 l2 = product (fun x y -> x,y) l1 l2
     end)
+    
+  let (and&) = combine_chop
```
Comment (c-cube) [inline src/core/CCList.ml:1813]:
```suggestion
  let (and&) = combine_shortest
```

File: src/core/CCList.mli [hunk 4]
```diff
@@ -138,6 +138,11 @@ val combine_gen : 'a list -> 'b list -> ('a * 'b) gen
     @since 1.2, but only
     @since 2.2 with labels *)
 
+val combine_chop : 'a list -> 'b list -> ('a * 'b) list
```
Comment (c-cube) [inline src/core/CCList.mli:141]:
```suggestion
val combine_shortest : 'a list -> 'b list -> ('a * 'b) list
```

File: src/core/CCList.mli [hunk 5]
```diff
@@ -138,6 +138,11 @@ val combine_gen : 'a list -> 'b list -> ('a * 'b) gen
...
+(** [combine [a1; …; am] [b1; …; bn]] is [[(a1,b1); …; (am,bm)]] if m <= n.
```
Comment (c-cube) [inline src/core/CCList.mli:142]:
```suggestion
(** [combine_shortest [a1; …; am] [b1; …; bn]] is [[(a1,b1); …; (am,bm)]] if m <= n.
```

File: src/core/CCList.mli [hunk 6]
```diff
@@ -814,6 +833,23 @@ module Infix : sig
   (** Let operators on OCaml >= 4.08.0, nothing otherwise
       @since 2.8 *)
   include CCShimsMkLet_.S with type 'a t_let := 'a list
+
+  val (and&) : 'a list -> 'b list -> ('a * 'b) list
+  (** [(and&)] is [combine_chop]. It allows to perform a synchronized product between two lists,
```
Comment (c-cube) [inline src/core/CCList.mli:838]:
```suggestion
  (** [(and&)] is [combine_shortest]. It allows to perform a synchronized product between two lists,
```

File: src/core/CCList.ml [hunk 7]
```diff
@@ -570,6 +572,29 @@ let combine_gen l1 l2 =
...
+
+(*$T
+  (combine_chop [] []) = []
+  (combine_chop [1] []) = []
+  (combine_chop [] [1]) = []
+  (combine_chop (1--1025) (1--1026)) = List.combine (1--1025) (1--1025)
+  (combine_chop (1--1026) (1--1025)) = List.combine (1--1025) (1--1025)
+  combine_chop [1;2;3] [3;2;1] = List.combine [1;2;3] [3;2;1]
+  combine_chop (1 -- 100_000) (1 -- 100_000) = List.combine (1 -- 100_000) (1 -- 100_000)
+  combine_chop (1 -- 100_001) (1 -- 100_000) = List.combine (1 -- 100_000) (1 -- 100_000)
```
Comment (c-cube) [inline src/core/CCList.ml:594]: I like that :+1:

**Review by c-cube — APPROVED** [review]
Excellent, will merge either when we apply suggestions, or if we pick a different name :-)

File: src/core/CCList.ml (hunk 2 above)
Comment (grayswandyr) [reply src/core/CCList.ml:582]: Suits me. Actually I copied your own code for `combine` ;-)

**Pull request merged**